api/app/src/main/kotlin/packit/service/RoleService.kt

```diff
@@ -0,0 +1,98 @@
+package packit.service
+
+import org.springframework.http.HttpStatus
+import org.springframework.security.core.GrantedAuthority
+import org.springframework.security.core.authority.SimpleGrantedAuthority
+import org.springframework.stereotype.Service
+import packit.exceptions.PackitException
+import packit.model.Role
+import packit.model.RolePermission
+import packit.repository.RoleRepository
+
+interface RoleService
+{
+    fun getUsernameRole(username: String): Role
+    fun getAdminRole(): Role
+    fun saveRole(roleName: String)
+    fun checkMatchingRoles(rolesToCheck: List<String>): List<Role>
+    fun getGrantedAuthorities(roles: List<Role>): MutableList<GrantedAuthority>
+}
+
+@Service
+class BaseRoleService(
+    private val roleRepository: RoleRepository
+) : RoleService
+{
+    override fun getUsernameRole(username: String): Role
+    {
+        val userRole = roleRepository.findByName(username)
+
+        if (userRole != null)
+        {
+            return userRole
+        }
+        return roleRepository.save(Role(name = username))
+    }
+
+    override fun getAdminRole(): Role
+    {
+        val userRole = roleRepository.findByName("ADMIN")
+
+        if (userRole != null)
+        {
+            return userRole
+        }
+        return roleRepository.save(Role(name = "ADMIN"))
+    }
+
+    override fun saveRole(roleName: String)
+    {
+        if (roleRepository.existsByName(roleName))
+        {
+            throw PackitException("roleAlreadyExists")
+        }
+        val role = Role(name = roleName)
+        roleRepository.save(role)
+    }
+
+    override fun checkMatchingRoles(rolesToCheck: List<String>): List<Role>
+    {
+        val allRoles = roleRepository.findAll()
+        val foundRoles = rolesToCheck.mapNotNull { name -> allRoles.find { it.name == name } }
```

Comment:
Again, is there not a nicer way via the ORM to pull out rows more efficiently that just getting everything?

Reply:
Yup fro sure,, i realised this in a subsequent PR --> https://github.com/mrc-ide/packit/pull/62/files#diff-7247283370c96b431d2ffb4dcc6b3bcf8e3d57e987e725bc675d9c80a61cd69f..